### ISSUE TITLE
Doesn't seem able to read flash BIOS on RX 560X

### ISSUE THREAD

**Issue by justinclift** [patrickschur/amdvbflash#2]
Just stumbled over this util while looking for some way to dump the VBIOS of my Radeon RX 560X from Linux, as I no longer have access to Windows.

However, it doesn't seem able to deal with the VBIOS on this graphics card. :frowning: 

Possibly weirdly, it's also not detecting the integrated AMD GPU on the system either. :wink:

### Standard approach

```
$ sudo ./amdvbflash -i
AMDVBFLASH version 4.69, Copyright (c) 2020 Advanced Micro Devices, Inc.


adapter seg  bn dn dID       asic           flash      romsize test    bios p/n    
======= ==== == == ==== =============== ============== ======= ==== ================
   0    0000 01 00 67EF Polaris11 (Slav R600 SPI         10000 fail       -
```

### Advanced

```
$ sudo ./amdvbflash -ai
AMDVBFLASH version 4.69, Copyright (c) 2020 Advanced Micro Devices, Inc.

Adapter  0    SEG=0000, BN=01, DN=00, PCIID=67EF1002, SSID=125D1025)
    Asic Family        :  Polaris11 (Slav
    Flash Type         :  R600 SPI    (64 KB)
    No VBIOS
```

### Attempting to dump the VBIOS

```
$ sudo ./amdvbflash -s 0 testfile.rom
AMDVBFLASH version 4.69, Copyright (c) 2020 Advanced Micro Devices, Inc.

Failed to read ROM
```

### Full system device list

This is an Acer Nitro 5 AN515-42 15.6" Full HD "Gaming" Laptop.  CPU is a Ryzen 7 2700U (ie with Radeon Vega Mobile)

```
$ sudo lspci
00:00.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Root Complex
00:00.2 IOMMU: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 IOMMU
00:01.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 17h (Models 00h-1fh) PCIe Dummy Host Bridge
00:01.1 PCI bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 PCIe GPP Bridge [6:0]
00:01.6 PCI bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 PCIe GPP Bridge [6:0]
00:01.7 PCI bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 PCIe GPP Bridge [6:0]
00:08.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Family 17h (Models 00h-1fh) PCIe Dummy Host Bridge
00:08.1 PCI bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Internal PCIe GPP Bridge 0 to Bus A
00:08.2 PCI bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Internal PCIe GPP Bridge 0 to Bus B
00:14.0 SMBus: Advanced Micro Devices, Inc. [AMD] FCH SMBus Controller (rev 61)
00:14.3 ISA bridge: Advanced Micro Devices, Inc. [AMD] FCH LPC Bridge (rev 51)
00:18.0 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 0
00:18.1 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 1
00:18.2 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 2
00:18.3 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 3
00:18.4 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 4
00:18.5 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 5
00:18.6 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 6
00:18.7 Host bridge: Advanced Micro Devices, Inc. [AMD] Raven/Raven2 Device 24: Function 7

*** RX 560X ***
01:00.0 Display controller: Advanced Micro Devices, Inc. [AMD/ATI] Baffin [Radeon RX 460/560D / Pro 450/455/460/555/555X/560/560X] (rev c0)
***************

02:00.0 Unassigned class [ff00]: Realtek Semiconductor Co., Ltd. RTL8411B PCI Express Card Reader (rev 01)
02:00.1 Ethernet controller: Realtek Semiconductor Co., Ltd. RTL8111/8168/8411 PCI Express Gigabit Ethernet Controller (rev 12)
03:00.0 Network controller: Qualcomm Atheros QCA6174 802.11ac Wireless Network Adapter (rev 32)
04:00.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Raven Ridge [Radeon Vega Series / Radeon Vega Mobile Series] (rev c3)
04:00.1 Audio device: Advanced Micro Devices, Inc. [AMD/ATI] Raven/Raven2/Fenghuang HDMI/DP Audio Controller
04:00.2 Encryption controller: Advanced Micro Devices, Inc. [AMD] Family 17h (Models 10h-1fh) Platform Security Processor
04:00.3 USB controller: Advanced Micro Devices, Inc. [AMD] Raven USB 3.1
04:00.4 USB controller: Advanced Micro Devices, Inc. [AMD] Raven USB 3.1
04:00.6 Audio device: Advanced Micro Devices, Inc. [AMD] Family 17h (Models 10h-1fh) HD Audio Controller
05:00.0 SATA controller: Advanced Micro Devices, Inc. [AMD] FCH SATA Controller [AHCI mode] (rev 61)
```

The discrete GPU itself seems fine:

```
$ sudo lspci -vv -s 01:00.0
01:00.0 Display controller: Advanced Micro Devices, Inc. [AMD/ATI] Baffin [Radeon RX 460/560D / Pro 450/455/460/555/555X/560/560X] (rev c0)
        Subsystem: Acer Incorporated [ALI] Device 125d
        Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0, Cache Line Size: 64 bytes
        Interrupt: pin A routed to IRQ 57
        Region 0: Memory at d0000000 (64-bit, prefetchable) [size=256M]
        Region 2: Memory at e0000000 (64-bit, prefetchable) [size=2M]
        Region 4: I/O ports at 3000 [size=256]
        Region 5: Memory at e0a00000 (32-bit, non-prefetchable) [size=256K]
        Expansion ROM at e0a40000 [disabled] [size=128K]
        Capabilities: [48] Vendor Specific Information: Len=08 <?>
        Capabilities: [50] Power Management version 3
                Flags: PMEClk- DSI- D1+ D2+ AuxCurrent=0mA PME(D0-,D1-,D2-,D3hot-,D3cold-)
                Status: D0 NoSoftRst- PME-Enable- DSel=0 DScale=0 PME-
        Capabilities: [58] Express (v2) Legacy Endpoint, MSI 00
                DevCap: MaxPayload 256 bytes, PhantFunc 0, Latency L0s <4us, L1 unlimited
                        ExtTag+ AttnBtn- AttnInd- PwrInd- RBE+ FLReset-
                DevCtl: Report errors: Correctable- Non-Fatal- Fatal- Unsupported-
                        RlxdOrd+ ExtTag+ PhantFunc- AuxPwr- NoSnoop+
                        MaxPayload 256 bytes, MaxReadReq 512 bytes
                DevSta: CorrErr- UncorrErr- FatalErr- UnsuppReq- AuxPwr- TransPend-
                LnkCap: Port #0, Speed 8GT/s, Width x8, ASPM L1, Exit Latency L1 <1us
                        ClockPM- Surprise- LLActRep- BwNot- ASPMOptComp+
                LnkCtl: ASPM L1 Enabled; RCB 64 bytes Disabled- CommClk+
                        ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
                LnkSta: Speed 2.5GT/s, Width x8, TrErr- Train- SlotClk+ DLActive- BWMgmt- ABWMgmt-
                DevCap2: Completion Timeout: Not Supported, TimeoutDis-, LTR+, OBFF Not Supported
                         AtomicOpsCap: 32bit- 64bit- 128bitCAS-
                DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis-, LTR-, OBFF Disabled
                         AtomicOpsCtl: ReqEn+
                LnkCtl2: Target Link Speed: 8GT/s, EnterCompliance- SpeedDis-
                         Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
                         Compliance De-emphasis: -6dB
                LnkSta2: Current De-emphasis Level: -3.5dB, EqualizationComplete+, EqualizationPhase1+
                         EqualizationPhase2+, EqualizationPhase3+, LinkEqualizationRequest-
        Capabilities: [a0] MSI: Enable+ Count=1/1 Maskable- 64bit+
                Address: 00000000fee01000  Data: 4022
        Capabilities: [100 v1] Vendor Specific Information: ID=0001 Rev=1 Len=010 <?>
        Capabilities: [270 v1] #19
        Capabilities: [320 v1] Latency Tolerance Reporting
                Max snoop latency: 0ns
                Max no snoop latency: 0ns
        Kernel driver in use: amdgpu
        Kernel modules: amdgpu
```

Not really sure if there's anything to be done, but if anyone has idea's I'm willing to experiment. :smile:

**Comment by justinclift** [comment]
As a data point, the same behaviour is shown when running the EFI version of amdvbflash (v2.93), from TechPowerUp:

https://www.techpowerup.com/download/ati-atiflash/

Same "No VBIOS" message. :wink:

**Comment by patrickschur** [comment]
You may have to unlock your ROM first.

```sh
sudo ./amdvbflash -unlockrom 0
```

Then try to dump the vBIOS again.

```sh
sudo ./amdvbflash -s 0 testfile.rom
```

> Possibly weirdly, it's also not detecting the integrated AMD GPU on the system either.

This tool doesn't support iGPUs. The vBIOS for iGPUs (and also for some dGPUs) is embedded in your UEFI. So have to dump your UEFI first and then extract the vBIOS. You may want to try the [VBiosFinder](https://github.com/coderobe/VBiosFinder) then.

**Comment by justinclift** [comment]
Thanks @patrickschur.  But still no go. :wink:

```
$ sudo ./amdvbflash -unlockrom 0
AMDVBFLASH version 4.69, Copyright (c) 2020 Advanced Micro Devices, Inc.

ROM Unlocked
$ 
```

```
$ sudo ./amdvbflash -s 0 testfile.rom
AMDVBFLASH version 4.69, Copyright (c) 2020 Advanced Micro Devices, Inc.

Failed to read ROM
```

```
$ sudo ./amdvbflash -ai
AMDVBFLASH version 4.69, Copyright (c) 2020 Advanced Micro Devices, Inc.

Adapter  0    SEG=0000, BN=01, DN=00, PCIID=67EF1002, SSID=125D1025)
    Asic Family        :  Polaris11 (Slav
    Flash Type         :  R600 SPI    (64 KB)
    No VBIOS
```

No worries about the iGPU, no real interest in it.  Am just attempting to dump the BIOS of the RX 560X, as there doesn't seem to be one available for download anywhere and I need one for attempting GPU passthrough with KVM.

But, super low priority, more from curiosity than anything.  Sounds like it's not a go-er at this stage. :wink:

**Comment by patrickschur** [comment]
What if you try to test access your ROM? I suspect this will fail to read the ROM, but it's worth a try.

```sh
sudo ./amdvbflash -t 0
```

Can you also give us some more information about your system? What Linux distro you are running, which kernel version and what laptop model you have? That would be great. :)

> as I no longer have access to Windows.

So you had access to Windows in the past? Did also flash your card in this time with a custom vBIOS?

**Comment by justinclift** [comment]
No worries, easily done. :smile:

Yep, you're right about it failing on that too:

```
$ sudo ./amdvbflash -t 0
AMDVBFLASH version 4.69, Copyright (c) 2020 Advanced Micro Devices, Inc.

Fail

   ERROR: 0FL02
```

Identical error code trying that after both `-lockrom` and -unlockrom` too.

System specs, I'm running OpenSUSE Leap 15.1, but with additional repos to provide [newer kernel](http://download.opensuse.org/repositories/Kernel:/stable/standard/) and [newer Mesa](https://download.opensuse.org/repositories/X11:/XOrg/openSUSE_Leap_15.1/).

The system was seriously unstable with the stock Mesa - full system lock ups a few times a day.  But with newer kernel and Mesa has been pretty decent.  Any kind of crash or lock up is rare (weeks apart, at least).

Kernel version at the moment is `5.6.7-3.1`.  Apparently it kernel 5.6.7, but with the OpenSUSE defaults applied.

Laptop is an Acer Nitro AN515-42.  It's a low/mid level gaming laptop. :smile:

Technically, it came with Win 10.  But I wiped the hdd and installed Linux on it right away, without even bothering to turn try out Win10.

At no point has a custom vBIOS been flashed onto this, it's just the same stock thing it came with. :smile:

---

Just to confirm the graphics card is actually working fine, I just booted Steam (on Linux) and ran Crysis Warhead (via Proton) for a bit on "Gamer" settings.  Worked fine, pretty sure there's no chance the iGPU was managing that. :grin:

**Comment by justinclift** [comment]
Just in case it's useful, these are the kernel boot parameters:

```
quiet rcu_nocbs=all rcu_nocb_poll pcie_aspm=off pci=pcie_bus_safe idle=nomwait \
intel_idle.max_cstate=0 processor.max_cstate=1 iommu=pt
```

**Comment by patrickschur** [comment]
**tldr**: I think your ROM is somehow r/w protected and you can't do much about it.

I've been dealing with this for over an hour now. But I couldn't find anything that would help you. The good news is you are not alone with this kind of error. The bad news is there is not really a fix for this error.

Some people say the card is somehow bricked (for me this doesn't make sense) and other people say, you have to connect the pin 1 and pin 8 on the back of your card together and then boot.
But I would strongly advise against that.

Things you could try:
- Test the tool on another Linux distro (e.g. Ubuntu)
- Use [Flashrom](https://software.opensuse.org/package/flashrom) to dump the vBIOS

**Comment by cd109876** [comment]
You often can't dump vbioses from laptops even in Windows, I found the vbios for the amd gpu in my laptop in the bios update file with VBiosFinder. No other method worked.

**Comment by justinclift** [comment]
No worries at all, thanks heaps for the suggestions and thoughts. :smile:

I'll attack the problem again later on with the Flashrom and VBiosFinder approach.  Hopefully one of them works for obtaining an RX 560X BIOS. :smile:
